### PR TITLE
Abstract away $in boilerplate when querying values

### PR DESCRIPTION
**Pull request by QuotableWater7**
**Summary**
Implementation of feature described in https://github.com/Automattic/mongoose/issues/4873

Normally, a query to find a key that matches multiple values would look like:
```js
SomeModel.find({ 
  _id: { 
    $in: [id1, id2, id3],
  },
})
```

With this change, the nested `$in` part would no longer be necessary.

```js
SomeModel.find({ _id: [id1, id2, id3] })
```

Motivation:  Cleaner code when writing queries.

**Test plan**
See the added tests in `cast.test.js`

### CODE REVIEW

File: lib/cast.js [hunk 1]
```diff
@@ -225,6 +225,13 @@ module.exports = function cast(schema, obj, options) {
             }
           }
         }
+      } else if (val.constructor.name === 'Array' && ['Buffer', 'Array'].indexOf(schematype.instance) === -1) {
+        var casted = [];
```
Comment (QuotableWater7) [inline lib/cast.js]: My preference would have been to put this logic into the individual `cast` methods for each type, but the problem is that in those `cast` methods, I have no idea if the value is actually supposed to be an array of values.  e.g. I could have a schema with:
```js
{
  someKey: { type: [Number] }
}
```
OR, I could have a schema with:
```js
{
  someKey: { type: Number } 
}
```

However, in `lib/schema/number.js`'s `cast` method, I don't know if the value is _supposed to be an array_, or if the value can simply be one of any number of values and thus can be converted to `$in` query.  

I would have passed another variable into the cast method at line `236` to contain the `schematype`, but the problem is having more than 1 argument will change how the method behaves entirely.

Long story short, I'm happy to hear possible improvements or a better location for this logic that can still pass the tests.

**Comment by vkarpov15** [comment]
This looks great :+1: I'll merge it when we release 4.8, which should be next week

File: lib/cast.js [hunk 2]
```diff
@@ -226,12 +226,9 @@ module.exports = function cast(schema, obj, options) {
           }
         }
       } else if (val.constructor.name === 'Array' && ['Buffer', 'Array'].indexOf(schematype.instance) === -1) {
-        var casted = [];
-        for (var valIndex = 0; valIndex < val.length; valIndex++) {
-          casted.push(schematype.castForQuery(val[valIndex]));
-        }
+        var castedValues = val.map(schematype.castForQuery.bind(schematype));
```
Comment (vkarpov15) [inline lib/cast.js]: Can you revert this change? `map()` is very neat and clean but much slower than for loop, and `.bind()` is extra slow.

Comment (QuotableWater7) [inline lib/cast.js]: @vkarpov15 I just reverted 👍
